### PR TITLE
Core Animation based camera transitions

### PR DESCRIPTION
**Pull request by frederoni**
WIP

This PR introduces Core Animation based camera transitions which can run concurrently and adaptively.

- [ ] Re-enable camera altitude based on road class
- [ ] Offset user location according to anchor point
- [ ] Improve dynamism of user course view positioning
- [x] Animate overview camera to fit bounds

@mapbox/navigation-ios 👀

### CODE REVIEW

**Comment by 1ec5** [comment]
Related: mapbox/mapbox-gl-native#8176 https://github.com/mapbox/mapbox-gl-native/issues/9808#issuecomment-325163145.

The caveat about animating across long distances still technically applies, but the nice thing about implementing this functionality in the navigation SDK is that we can assume that the animation typically occurs only over short distances.

File: MapboxNavigation/NavigationMapView.swift [hunk 1]
```diff
@@ -226,6 +227,9 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
     }
     
     fileprivate func commonInit() {
+        navigationCamera = NavigationCamera(self)
```
Comment (JThramer) [inline MapboxNavigation/NavigationMapView.swift]: `NavigationCamera(self)` isn't very descriptive, and misleads me into thinking that this is something functional.

How about `NavigationCamera(mapView: MGLMapView)`?

File: MapboxNavigation/NavigationMapView.swift [hunk 2]
```diff
@@ -1104,19 +1108,39 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
...
     @objc public func setOverheadCameraView(from userLocation: CLLocationCoordinate2D, along coordinates: [CLLocationCoordinate2D], for bounds: UIEdgeInsets) {
-        let slicedLine = Polyline(coordinates).sliced(from: userLocation).coordinates
-        let line = MGLPolyline(coordinates: slicedLine, count: UInt(slicedLine.count))
-        
-        tracksUserCourse = false
-        let camera = self.camera
-        camera.pitch = 0
-        camera.heading = 0
-        self.camera = camera
+//        let slicedLine = Polyline(coordinates).sliced(from: userLocation).coordinates
```
Comment (JThramer) [inline MapboxNavigation/NavigationMapView.swift]: Don't forget the comments

File: MapboxNavigation/NavigationMapView.swift [hunk 3]
```diff
@@ -1104,19 +1108,39 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
...
+//        // Don't keep zooming in
+//        guard line.overlayBounds.ne.distance(to: line.overlayBounds.sw) > NavigationMapViewMinimumDistanceForOverheadZooming else { return }
+//
+//        setVisibleCoordinateBounds(line.overlayBounds, edgePadding: bounds, animated: true)
+//        self.animView?.percent = 0
+//        UIView.animate(withDuration: SWViewController.kAnimDuration) {
+//            self.animView?.percent = 100.0
+//        }
+        
+        self.navigationCamera.altitude = 100
```
Comment (JThramer) [inline MapboxNavigation/NavigationMapView.swift]: Could we move all of these inline magic values into constants?

File: MapboxNavigation/NavigationCameraLayer.swift [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            course = layer.course
+        }
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+    }
+    
+    fileprivate class func isCustomAnimationKey(key: String) -> Bool {
+        return !(CustomAnimationKey(rawValue: key) == nil)
```
[the remark this answers is not in the document]
Comment (JThramer) [reply MapboxNavigation/NavigationCameraLayer.swift]: (or maybe `maps(to value: RawValue) -> Bool`)

File: MapboxNavigation/NavigationMapView.swift [hunk 5]
```diff
@@ -1104,19 +1108,39 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
...
-        // Don't keep zooming in
-        guard line.overlayBounds.ne.distance(to: line.overlayBounds.sw) > NavigationMapViewMinimumDistanceForOverheadZooming else { return }
+        UIView.animate(withDuration: 1, delay: 1, options: [], animations: {
+            self.navigationCamera.pitch = 0
+        }, completion: nil)
         
-        setVisibleCoordinateBounds(line.overlayBounds, edgePadding: bounds, animated: true)
+        UIView.animate(withDuration: 1, delay: 2, options: [], animations: {
+            self.navigationCamera.course = 0
+        }, completion: nil)
```
Comment (bsudekum) [inline MapboxNavigation/NavigationMapView.swift]: How are you figuring out the proper center coordinate here?

Comment (frederoni) [reply MapboxNavigation/NavigationMapView.swift]: Center coordinate was added in https://github.com/mapbox/mapbox-navigation-ios/pull/1240/commits/8d33ef76bb9340d7c7ec2d70c2c04cfee2d3b539. Next up is visible bounds.

**Comment by frederoni** [comment]
Getting closer but still a lot of caveats.

File: MapboxNavigation/RouteMapViewController.swift [hunk 6]
```diff
@@ -340,25 +343,25 @@ class RouteMapViewController: UIViewController {
...
-        //If the user is on a motorway, not exiting, and their segment is sufficently long, the map should zoom out to the motorway altitude.
-        //otherwise, zoom in if it's the last instruction on the step.
-        let currentStepIsMotorway = currentStep.isMotorway
-        let nextStepIsMotorway = upComingStep?.isMotorway ?? false
-        if currentStepIsMotorway, nextStepIsMotorway, isLongRoad {
-            setCamera(altitude: zoomOutAltitude)
-        } else if currentInstruction == currentStep.lastInstruction {
-            setCamera(altitude: defaultAltitude)
-        }
+//
```
Comment (frederoni) [inline MapboxNavigation/RouteMapViewController.swift]: note to self: move this into `navigationCamera.setCourseTracking`

File: MapboxNavigation/NavigationMapView.swift [hunk 7]
```diff
@@ -338,15 +339,9 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
...
-            let point = userAnchorPoint
-            let padding = UIEdgeInsets(top: point.y, left: point.x, bottom: bounds.height - point.y, right: bounds.width - point.x)
-            let newCamera = MGLMapCamera(lookingAtCenter: location.coordinate, fromDistance: altitude, pitch: 45, heading: location.course)
-            let function: CAMediaTimingFunction? = animated ? CAMediaTimingFunction(name: kCAMediaTimingFunctionLinear) : nil
-            setCamera(newCamera, withDuration: duration, animationTimingFunction: function, edgePadding: padding, completionHandler: {
-                UIView.animate(withDuration: duration, delay: 0, options: [.curveLinear, .beginFromCurrentState], animations: {
-                    self.userCourseView?.center = self.convert(location.coordinate, toPointTo: self)
-                }, completion: nil)
-            })
+            // TODO: offset centerCoordinate according to `userAnchorPoint`
```
Comment (vincethecoder) [inline MapboxNavigation/NavigationMapView.swift]: Should we leave TODO comments in the public sdk?

File: MapboxNavigation/NavigationMapView.swift [hunk 8]
```diff
@@ -1101,20 +1096,28 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
...
+    }
+    
+    func transitionToCourseTracking(completion: NavigationCamera.CompletionHandler = nil) {
+        guard let location = userLocationForCourseTracking else {
+            completion?(false)
+            return
+        }
         
-        setVisibleCoordinateBounds(line.overlayBounds, edgePadding: bounds, animated: true)
+        navigationCamera.transitionToCourseTracking(duration: 2, // TODO: Set duration based on travel distance
```
Comment (vincethecoder) [inline MapboxNavigation/NavigationMapView.swift]: Maybe we should convert this `TODO` into an issue ticket to be addressed later.

File: MapboxNavigation/NavigationMapView.swift (hunk 7 above)
Comment (frederoni) [reply MapboxNavigation/NavigationMapView.swift]: Nah, I add these to remind myself what's left to do before final review and merge.

**Comment by frederoni** [comment]
CI finally caught the pod integration bug introduced in https://github.com/mapbox/mapbox-navigation-ios/pull/1240/commits/5d44763a8405130a62e9bb7d5341c428aced6a10 fixed in https://github.com/mapbox/mapbox-navigation-ios/pull/1240/commits/1f7036f1cf9d64693fb5cbe3e9a2eaeb5d84a02b

**Comment by bsudekum** [comment]
@frederoni what's the latest on this PR? Do you think we should include this in v0.17.0 [milestone](https://github.com/mapbox/mapbox-navigation-ios/milestone/17)?

File: MapboxNavigation/NavigationMapView.swift [hunk 9]
```diff
@@ -1101,20 +1097,28 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
...
-        camera.heading = 0
-        self.camera = camera
-        
-        // Don't keep zooming in
-        guard line.overlayBounds.ne.distance(to: line.overlayBounds.sw) > NavigationMapViewMinimumDistanceForOverheadZooming else { return }
+        let slicedLine = Polyline(coordinates).sliced(from: userLocation).coordinates
+        navigationCamera.transitionToOverview(coordinates: slicedLine, edgePadding: edgePadding)
+    }
+    
+    func updateAltitude(_ altitude: CLLocationDistance) {
```
Comment (leviathan) [inline MapboxNavigation/NavigationMapView.swift]: Would you mind exposing this method as `public` ?!
I would like to update the camera `altitude` from the outside, e.g. using the `NavigationMapView` as a single entity.

**Pull request closed without merging**